HARK/ConsumptionSaving/ConsIndShockModel_Both.py

```diff
@@ -0,0 +1,330 @@
+
+    Parameters
+    ----------
+    soln : agent_stage_solution
+        A solution with an atached Modl object.
+
+    transition_name : str
+        Name of the step in Pars.transitions_possible to execute
+
+    Returns
+    -------
+    None.
+
+    """
+    transitions_possible = soln.Bilt.transitions_possible
+    transition = transitions_possible[transition_name]
+    transiter = {}  # It's the actor that actually does the job
+    transiter['compiled'] = {}
+    transiter['raw_text'] = {}
+    transiter['last_val'] = {}
+
+    for eqn_name in transition:
+        transiter['raw_text'].update({eqn_name: transition[eqn_name]})
+        tree = parse(transition[eqn_name], mode='exec')
+        compiled = compile(tree, filename="<ast>", mode='exec')
+        transiter['compiled'].update({eqn_name: compiled})
+
+    soln.Modl.Transitions[transition_name] = transiter
+    return soln
+
+
+def def_utility_CRRA(soln, CRRA):
+    """
+    Define CRRA utility function and its relatives (derivatives, inverses).
+
+    Saves them as attributes of self for other methods to use.
+
+    Parameters
+    ----------
+    soln : ConsumerSolutionOneStateCRRA
+
+    Returns
+    -------
+    none
+    """
+    Bilt, Pars, Modl = soln.Bilt, soln.Pars, soln.Modl
+    Info = Modl.Info = {**Bilt.__dict__, **Pars.__dict__}
+
+    Modl.Rewards = SimpleNamespace()
+
+    Modl.Rewards.raw_text = {}
+    Modl.Rewards.eqns = {}
+    Modl.Rewards.vals = {}
+
+    # Add required funcs to Modl.Info
+    for func in {'CRRAutility', 'CRRAutilityP', 'CRRAutilityPP',
+                 'CRRAutility_inv', 'CRRAutility_invP', 'CRRAutilityP_inv',
+                 'CRRAutilityP_invP'}:
+        Info[func] = globals()[func]
+
+    # Hard-wire the passed CRRA into the utility function and its progeny
+    eqns_source = {
+        'u_D0':
+            'u = lambda c: CRRAutility(c,' + str(CRRA) + ')',
+        'u_D1':
+            'u.dc = lambda c: CRRAutilityP(c, ' + str(CRRA) + ' )',
+        'u_D2':
+            'u.dc.dc = lambda c: CRRAutilityPP(c, ' + str(CRRA) + ')',
+        'uNvrs_D0':
+            'u.Nvrs = lambda u: CRRAutility_inv(u, ' + str(CRRA) + ')',
+        'uNvrs_D1':
+            'u.Nvrs.du = lambda u: CRRAutility_invP(u, ' + str(CRRA) + ' )',
+        'u_D1_Nvrs':
+            'u.dc.Nvrs = lambda uP: CRRAutilityP_inv(uP, ' + str(CRRA) + ')',
+        'u_D1_Nvrs_D1':
+            'u.dc.Nvrs.du = lambda uP: CRRAutilityP_invP(uP, ' + str(CRRA) + ')',
+    }
+
+    # Put the utility function in the "rewards" part of the Modl object
+    for eqn_name in eqns_source.keys():
+        tree = parse(eqns_source[eqn_name], mode='exec')
+        code = compile(tree, filename="<ast>", mode='exec')
+        Modl.Rewards.eqns.update({eqn_name: code})
+        exec(code, {**globals(), **Info}, Modl.Rewards.vals)
+
+        Modl.Rewards.raw_text = eqns_source
+
+    Bilt.__dict__.update({k: v for k, v in Modl.Rewards.vals.items()})
+
+    return soln
```

Comment:
I am confused why a utility function (which knows its derivatives, which is an awesome addition!) returns a solution object.

Seems like a mathematical function is more basic than a solution.

Reply:
This was part of a pattern of building up the contents of the solution object with standalone commands like

soln = self.def_utility_CRRA()
soln = self.def_value_funcs()

Probably would be an improvement to have self.def_utility_CRRA() return a 'utility' object, def_value_funcs a vFunc, etc.  Then the calling code could become:

soln.u = self.def_utility_CRRA()
soln.vFunc = self.def_value_funcs()

The way these kinds of things are done right now is rather haphazard and inconsistent.  An improvement in the code would be to systematize and regularlize these kinds of things.